DM-32940: Add new Butler methods for checking existence.

timj opened the pull request:
These replace `Butler.datasetExists()`.

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

timj committed:
Add new Butler.exists() method

This replace Butler.datasetExists

timj committed:
Add LimitedButler.stored() API

timj committed:
Add butler.exists_many

timj committed:
Add qbb stored() and stored_many()

timj committed:
WIP: Switch to .exists()

timj committed:
Change datasetExists to use new API in tests

timj committed:
Remove getCheckedId usage

timj committed:
Adjust stacklevel for new warnings

This makes them appear to come from user code rather than daf_butler
internals.

timj committed:
Rename test packer to stop pytest complaining

Pytest really wants to test every class that starts with the
name Test.

timj committed:
Fix new warnings from test code

timj committed:
Add tests for Butler.exists and Butler.exists_many

timj committed:
Add stored_many test to QBB test

timj commented:
@TallJimbo can you please take a look at these changes ahead of the formal Jira review? I have to change a bunch of other packages to use the new API but if you think I've done the API incorrectly I'd rather know ahead of time. This PR is essentially complete with a caveat that it seems to me that `QuantumBackedButler` could implement `exists` where `RegistryKnows` would be true if the ref is in the graph at all.

timj commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -111,6 +114,19 @@
 log = getLogger(__name__)
 
 
+class DatasetExistence(Flag):
```
This will likely have to move to a separate file (`_enums.py`?) if we are going to add `QuantumBackedButler.exists()`.

TallJimbo replied:
:+1:, but probably better to just make it `_dataset_existence.py`, unless we already have other enums with approximately the same downstream consumers we want to bundle up.

timj commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -111,6 +114,19 @@
 
+class DatasetExistence(Flag):
+    Unknown = 0
+    RegistryKnows = auto()
```
I am not attached to these names if people have opinions.

TallJimbo replied:
I think they should be ALL_CAPS, right?  Or have we already been inconsistent in that sense about enum values?  I know `CollectionType` is all caps.

timj replied:
All caps seemed overkill and shouty for `ARTIFACT_NOT_CHECKED` but that's fine but if that's the policy for enums.

timj commented on python/lsst/daf/butler/_limited_butler.py:
```diff
@@ -273,6 +276,52 @@ def getDeferred(
+    def stored_many(
+        self,
+        refs: Iterable[DatasetRef],
+        artifact_existence: Optional[dict[ResourcePath, bool]] = None,
```
I am torn as to whether `artifact_existence` should be a parameter here. It's useful as an opaque cache if people know they are going to call it many times but only if they also know that nothing is actively writing files between calls to it (which `transfer_from` does assume). I think I'm talking myself out of having it here as being too dangerous.

timj replied:
I think I've talked myself out of forwarding `artifact_existence` for the above reasons.

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1597,6 +1613,142 @@ def retrieveArtifacts(
+        self,
+        datasetRefOrType: DatasetRef | DatasetType | str,
+        /,
+        dataId: DataId | None = None,
```
How about `data_id` here instead?  I know the analogous methods have `dataId`, but I think new ones are an opportunity to move us along the path to snake_case consistency, such that some day we'll just just have a few camelCase ones left and then it'll be worth it to go after changing them directly.

timj replied:
I was a bit worried that people are so used to say `dataId=` in `butler.get` and `butler.getURI` that they will do the same thing for `butler.exists`.

timj replied:
Okay. I switched over.

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1597,6 +1613,142 @@ def retrieveArtifacts(
+        -------
+        existence : `DatasetExistence`
+            Object indicating whether the dataset is known to registry and
+            datastore. Evaluates to `True` if the dataset is present is known
```
```suggestion
            datastore. Evaluates to `True` if the dataset is present and known
```

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1597,6 +1613,142 @@ def retrieveArtifacts(
+
+        if isinstance(datasetRefOrType, DatasetRef):
+            registry_ref = self.registry.getDataset(datasetRefOrType.id)
+            if registry_ref is not None:
```
We probably want a check here to see if the data ID and dataset type we got are the same (or compatible with) what we were given.  I think that not being the case should be an exception, but we also could make "present in registry, but not as defined" another enum flag.

timj replied:
I was not expecting a situation where a UUID could match but somehow be associated with a completely different ref. Doesn't that seem unlikely? I can add the check of dataId and datasetType if you think it's possible. I really need a `DatasetRef.is_compatible_with` call or something (since `__eq__` does not check for storage class compatibility)

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1597,6 +1613,142 @@ def retrieveArtifacts(
+            When `DatasetRef` the `dataId` should be `None`.
+            Otherwise the `DatasetType` or name thereof.
+            If a resolved `DatasetRef`, the associated dataset
+            is returned directly without additional querying.
```
I'd consider the `getDataset` call "additional querying", since it still involves hitting the database (same comment for `exists_many`).

timj replied:
Yes. That's a copy and paste problem from `butler.get`

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1597,6 +1613,142 @@ def retrieveArtifacts(
+            Mapping from the original dataset to an enum indicating the
+            status of the dataset in registry and datastore.
+            Each value evaluates to `True` if the dataset is present is known
+            to both.
```
Should state whether it's guaranteed that all given refs will be present as keys, and if so, whether they'll be in the same order.

We also need to be really careful that all returned refs are identical to the given ones, not e.g. be mostly the same but have a compatible storage class.  I think this method is safe in that regard only if the datastore methods it calls are safe in that regard.

timj replied:
`datastore.mexists()` does guarantee that the keys are the refs you gave it even if the refs had different storage classes. I've updated the docs.

TallJimbo commented on python/lsst/daf/butler/_limited_butler.py:
```diff
@@ -273,6 +276,52 @@ def getDeferred(
         return DeferredDatasetHandle(butler=self, ref=ref, parameters=parameters, storageClass=storageClass)
 
+    def stored(self, ref: DatasetRef) -> bool:
+        """Indicate whether the dataset is present in the Datastore.
```
Should be more precise about what this means, e.g. "whether the dataset's artifacts are all present in the Datastore".

TallJimbo commented on python/lsst/daf/butler/_limited_butler.py:
```diff
@@ -273,6 +276,52 @@ def getDeferred(
+            this method with details of all artifacts tested. Can be `None`
+            if the caller is not interested but can be an efficiency gain
+            if you know that this API is going to called repeatedly with
+            possibly duplicate refs.
```
Is the idea that anything in this `dict` already is not checked?  Docs may need some clarification.

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -111,6 +114,19 @@
+    ArtifactExists = auto()
+    ArtifactNotChecked = auto()
+    Exists = RegistryKnows | DatastoreKnows | ArtifactExists
+    PartialCheck = RegistryKnows | DatastoreKnows | ArtifactNotChecked
```
I'd like to handle `ArtifactNotChecked` differently somehow; we also don't check artifact existence when `~RegistryKnows` or `~DatastoreKnows`, so we should at least rename it to something like `ArtifactAssumed`.  But I also think it might be better to make `DatastoreKnows & ~ArtifactExists` back into an error, since that's really not a state the repository should ever be in (and that'll be even more seriously the case after DMTN-249).  We could then drop the `ArtifactNotChecked` flag, with the default being that artifact existence is assumed if `DatastoreKnows` and verified if `ArtifactExists`.

We could also consider dropping `ArtifactExists`, too, since its value would then be wholly deterministic from `DatastoreKnows & full_check`, and the caller could be expected to know what they passed as `full_check`, but I have no objection to keeping it.

If we keep `ArtifactExists`, I think we should also rename `PartialCheck` (` = RegistryKnows | DatastoreKnows`) to something like `FastExists` and rename `Exists` to `ExhaustiveExists` so the most default "does exist" result doesn't imply that the user should have done more.

I think we could accomplish all that in terms of flags, but here's a regular enum version that may be worth considering:
```
class DatasetExistence(Enum):
    UNRECOGNIZED = auto()   # No matching dataset in Registry.  Datastore not checked.
    RECORDED = auto()   # In Registry, but not in any Datastore (per datastore records)
    COMPLETE = auto()  # In Registry and Datastore (per datastore records)
    EXHAUSTIVE = auto()  # In Registry and Datastore, artifact verified

    def __bool__(self) -> bool:
        return self is self.COMPLETE or self is self.EXHAUSTIVE
```
The advantage of this is that it can only represent states that make sense (e.g. no `~RegistryKnows & DatastoreKnows` or `ArtifactExists & ArtifactNotChecked`).  The disadvantage is that you can't just OR in new flags to "upgrade" the state.

timj replied:
I used `ArtifactNotChecked` because it seemed to simplify the bool logic (and I had to check `.value` because otherwise I ended up with recursion errors) and allowed the "truth" checks to work consistently -- I wasn't expecting it to be a public flag that people would check against but was an internal assertion that everything was okay and had been checked.

timj replied:
I also wasn't expecting PartialCheck to be public. It was also there to simplify the boolean. I was assuming that people would either just use truthiness or be asking for RegistryKnows or DatastoreKnows explicitly.

timj replied:
My other comment is that I much prefer compound enum states. RECORDED meaning in registry but not in datastore makes it harder to code because it means that you need to keep track of the registry state and then later decide whether to override with datastore state. Whereas OR-ing together as you go and then using combination flags for the public information is easier to calculate at the end.

timj replied:
I've rewritten and renamed the Flags and hidden flag names for the artifact existence work well.

TallJimbo reviewed (commented):
My preferred interpretation of the `RegistryKnows` state is that it really means "in the central database for this data repository", not "in whatever is providing metadata about datasets for this butler client".  I suppose we could extend `exists` to return something about both, and then `LimitedButler` could have a version of `exists` that can takes `DatasetRef` only) and can say yes to the latter while always saying "I don't know" for the former.  But that's a bunch of of new `DatastoreExistence` states that I'm not sure carry their weight.

timj committed:
Fix some docstring problems

timj committed:
Rename DatasetExistence flags

Use upper case and now use hidden flag names for whether the
artifact exists or not (since people shouldn't really need
to know about the difference).

timj committed:
Fix butler.exists docstring

timj committed:
Remove artifact_existence cache and tweak docstrings for stored

timj committed:
Use modern snake case for Butler.exists parameters

Since this is a new API.

timj committed:
Add DatasetRef.is_compatible_with

timj committed:
Doc string clarification

timj committed:
Check the registry ref against given ref in Butler.exists

This is for the case where the UUIDs match but something
else is different such as storage class.

timj committed:
Mark Butler._exists_many as private

We are not entirely sure that this is the right API given the
possibilities that refs can be given that were crafted externally
but still look like they share the same ID (storage class overrides
notwithstanding). The API would be cleaner if we used iterable[UUID]
instead but that needs some thought. For now mark it private but
still test it.

timj committed:
Add news fragment

kfindeisen commented on python/lsst/daf/butler/registry/queries/_query_backend.py:
```diff
@@ -624,8 +623,7 @@ def resolve_governor_constraints(
         -------
-        resolved : `Mapping` [ `str`,  [ `~collections.abc.Set`
-                [ `str` ] ] ]
+        resolved : `Mapping` [ `str`,  [ `~collections.abc.Set` [ `str` ] ] ]
```
FTR, the [Python style guide](https://developer.lsst.io/python/numpydoc.html#py-docstring-parameter-types-dict) only requires spaces _before_ brackets, not inside them. The extra space adds up for highly nested types, and AFAIK it has no effect on the final HTML.

timj replied:
I'm dreaming of a time when I don't need to put any of this in because the type annotations will be used directly.

kfindeisen replied:
Personally, I hope that doesn't happen. Some of the broken links on #846 were the result of trying to use the exact same notation as the annotations, even when Sphinx couldn't understand them. In a few cases (e.g., "any collection expression"), the human-readable type is _much_ simpler than its machine-readable equivalent (`str | Pattern | Iterable[str | Pattern] | ellipsis | CollectionWildcard | None`).

timj commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1597,6 +1599,161 @@ def retrieveArtifacts(
+
+        return existence
+
+    @deprecated(
```
I think I will have to remove this message for the moment since we have not filed an RFC formally deprecating the interface.

andy-slac replied:
Probably makes sense to add it on an implementation ticket for that RFC.

timj replied:
[RFC-933](https://jira.lsstcorp.org/browse/RFC-933) has been approved.

andy-slac commented on python/lsst/daf/butler/_dataset_existence.py:
```diff
@@ -0,0 +1,82 @@
+
+class DatasetExistence(Flag):
+    """Flags representing the different states that a dataset can have
+    in a Butler repository."""
```
Maybe expand documentation a little bit to help people who are not very familiar with the combination of flags. Generated docs do not show that `KNOWN = RECORDED | DATASTORE | _ASSUMED`, it just shows that `KNOWN = 11` which could be confusing. And maybe explain logic behind boolean conversion, docs for `__bool__` method do not appear in the generated output.

timj replied:
The doc strings do explain which flags evaluate to true and which evaluate to false but I've added some text.

andy-slac commented on python/lsst/daf/butler/_dataset_existence.py:
```diff
@@ -0,0 +1,82 @@
+            Evaluates to `True` if the flags evaluate to either ``KNOWN``
+            or ``VERIFIED``.
+        """
+        return self.value == self.KNOWN.value or self.value == self.VERIFIED.value
```
You could probably write `self is self.KNOWN or self is self.VERIFIED`?

timj replied:
I will try that again. When I did `self == self.KNOWN` I got a recursion exception but I had not tried using `is`.

andy-slac commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1597,6 +1599,161 @@ def retrieveArtifacts(
+        if full_check:
+            if self.datastore.exists(ref):
+                existence |= DatasetExistence._ARTIFACT
+        elif existence != DatasetExistence.UNRECOGNIZED:
+            # Do not add this flag if we have no other idea about a dataset.
+            existence |= DatasetExistence._ASSUMED
```
Should this check be for `existence | DatasetExistence.DATASTORE`, does it make sense to have `_ASSUMED` without `DATASTORE`? Maybe whole `if full_check:` wants to be nested in `if self.datastore.knows(ref):`?

timj replied:
It is possible for the file to exist but for datastore not to know about it. That's the reason we have to two independent states. This is how execution butler runs and how `trustGetRequest` works. `full_check` therefore can't be nested inside `knows`. In a "proper" datastore the `datastore.exists` will check if it knows first and return False regardless.

TallJimbo replied:
I may have misunderstood what happens in trust mode.  Does that never return DATASET_KNOWS or ARTIFACT_ASSUMED?

I guess that makes sense but I worry that it exposes an implementation detail (whether the datastore is in trust mode).

Of course, that may not matter if we can get to the point where trust mode is only ever paired with QBB, not a full Butler, after execution butler is gone.  I'm hoping we can.

timj replied:
`_ASSUMED` is a private flag entirely there to make things obvious internally. It's not a public state. In execution butler it will return False because it really doesn't know about the ref and full check will indicate that the artifact does exist. You will see extensive tests exploring all the different states. If `datastore.trustGetRequest` is something that `Butler.exists` is aware of then we could make it look on the file system regardless of `full_check` and add `DATASTORE` flag if the artifact exists.

TallJimbo replied:
I think I've convinced myself I don't care what happens with a full Butler in `trustGetRequest` mode, since I just want to make that combination go away entirely instead.

andy-slac commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1597,6 +1599,161 @@ def retrieveArtifacts(
+    def _exists_many(
+        self,
+        refs: Iterable[DatasetRef],
+        /,
+        *,
```
Do you need `/, *,` if there is nothing in between?

timj replied:
I'm not entirely certain. I was giving myself flexibility but this is a private API at the moment so we can do whatever here.

andy-slac commented on python/lsst/daf/butler/_quantum_backed.py:
```diff
@@ -447,15 +447,40 @@ def getDeferred(
+    )
+    def datasetExistsDirect(self, ref: DatasetRef) -> bool:
+        # Docstring inherited.
+        return self.stored(ref)
```
It looks the same as base class implementation, could be dropped?

andy-slac commented on python/lsst/daf/butler/core/datasets/ref.py:
```diff
@@ -628,6 +628,56 @@ def overrideStorageClass(self, storageClass: str | StorageClass) -> DatasetRef:
+        """
+        if not isinstance(ref, DatasetRef):
+            # mypy thinks this is impossible but we need to prevent
+            # an AttributeError if possible.
```
Not clear where the AttributeError comes from?

timj replied:
It will come from the `ref.id` below if you pass the wrong thing here. mypy never expects you to pass the wrong thing here so is convinced it can't happen.

andy-slac replied:
But this is a completely new method, don't we control what is being passed to it? And if someone passes and incorrect type then they probably deserve AttributeError, not? Otherwise if we allow other types maybe just declare argument as `ref: object`?

timj replied:
I was being defensive since it felt to me that this was a lot like `__eq__` which does have to worry about any random thing being given to it. If you feel that AttributeError is fine if someone passes in a non-ref I'll change it.

andy-slac replied:
I think that checking a non-ref to be compatible with a ref is a meaningless operation and no one is going to ask those questions, so an exception should be fine here. And Python 4 should probably raise TypeError if an actual argument type is different from annotation 🙂

timj replied:
I've removed it, it did remind me that `__eq__` does not check that the `run` matches, only the ID, dataset type, and dataID...

andy-slac commented on tests/test_butler.py:
```diff
@@ -342,8 +343,7 @@ def runPutGetTest(self, storageClass: StorageClass, datasetTypeName: str) -> But
                 # Now remove the dataset completely.
                 butler.pruneDatasets([ref], purge=True, unstore=True)
                 # Lookup with original args should still fail.
-                with self.assertRaises(LookupError):
-                    butler.datasetExists(*args, collections=this_run)
```
Maybe keep tests for `datasetExists` for now until it's actually deprecated?

andy-slac reviewed (approved):
Looks good, few minor questions.

andy-slac commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1597,6 +1599,161 @@ def retrieveArtifacts(
+            to query an external system. If `False` registry and datastore
+            will solely be asked if they know about the dataset but no
+            check for the artifact will be performed.
+        collections : Any, optional
+            Collections to be searched, overriding ``self.collections``.
+            Can be any of the types supported by the ``collections`` argument
+            to butler construction.
```
One more thought - it looks like `collections` is not used when `ref` is given, meaning that we'll find a ref even if it's not in any of the `collections`. Maybe clarify that in the docstring to avoid surprises for users who may think that `collections` could be used as restriction even if dataset ref is provided?

timj replied:
Right. We do warn in `Butler.get()` now if you pass ref and collections.

timj committed:
Expand docs for DatasetExistence

timj committed:
Remove datasetExistsDirect from QBB

It now matches the base class version.

timj committed:
Remove the check for DatasetRef in is_compatible_with

If someone does not pass a ref in then let it break.

timj committed:
Add warning if Butler.exists called with ref + dataID + collection
